### ISSUE TITLE
交叉熵损失 API 设计

### ISSUE THREAD

**Issue by dangkai4u** [Oneflow-Inc/oneflow#7167]
在oneflow里，交叉熵损失有以下几种：
- binary_cross_entropy_loss
- binary_cross_entropy_with_logits_loss
- sparse_cross_entropy
- distributed_sparse_cross_entropy
- cross_entropy
- sparse_softmax_cross_entropy
- softmax_cross_entropy

在pytorch里，交叉熵损失有以下几种：
- binary_cross_entropy 
- binary_cross_entropy_with_logits 
- cross_entropy 

由此可见，oneflow中交叉熵损失存在API冗余，重复，容易让用户疑惑，因此，这里应该精简一下。除此之外，label smoothing是一个非常有效的避免过拟合技巧，经常和交叉熵损失结合，目前pytorch的cross_entropy也支持设置label smoothing，如果可以支持，开发模型时，就可以直接调用内置函数，而非重新实现损失函数。

**Comment by dangkai4u** [comment]
## pytorch交叉熵损失设计：
### BCELoss & BCEWithLogitsLoss
- 这两个模块都是实现了二元交叉熵损失计算。`BCELoss`的输入为使用`sigmoid`归一化后的数值，`BCEWithLogitsLoss`的输入为`logits`，后者相比于前者，由于使用`log-sum-exp`计算，数值更加稳定。
- 参数设计：
    ```python
    torch.nn.BCELoss(weight=None, size_average=None, reduce=None, reduction='mean')
    torch.nn.BCEWithLogitsLoss(weight=None, size_average=None, reduce=None, reduction='mean', pos_weight=None)
    ```
    - `weight`：一个标识每个样本权重的张量，用于对样本重加权。为了实现这个功能，pytorch实现了一个`_WeightedLoss`父类，其它需要该功能的损失函数都继承它。
    - `size_average`：被废弃，不需要实现。
    - `reduce`：被废弃，不需要实现。
    - `reduction`：类型是字符串，“mean”表示对结果求平均，“sum”表示对结果求和，“none”表示不进行规约操作。`size_average`和`reduce`功能与它相同，而它更全面，因此那两个参数废弃了。
    - `pos_weight`：`BCEWithLogitsLoss`特有的，用于对正样本加权。我没有使用过，感觉和`weight`功能重叠，可以通过`weight`代替。

### CrossEntropyLoss & NLLLoss
- 这两个为一组，都实现了交叉熵损失的计算。`CrossEntropyLoss`的输入为`logits`，`NLLLoss`的输入为`logits`经过`log_softmax`后的结果，`CrossEntropyLoss`更具有数值稳定性。
- 参数设计：
    ```python
    torch.nn.CrossEntropyLoss(weight=None, size_average=None, ignore_index=- 100, reduce=None, reduction='mean', label_smoothing=0.0)
    torch.nn.NLLLoss(weight=None, size_average=None, ignore_index=- 100, reduce=None, reduction='mean')
    ```
    - `weight`、`size_average`、`reduce`、`reduction`：同上
    - `ignore_index`：拿文本生成任务举例，训练时会对文本填充<pad>，这些pad字符也会生成logits，但我们不希望它计算损失，因此通过指定`ignore_index=pad_idx`，忽略这些字符的影响。这是一个比较常见且实用的功能，应当予以实现。`ignore_index`是一个整数，不一定属于label空间，可以为任意数，包括负数。如果`reduction`为mean时，`ignore_index`所指的样本不应该计算在内（不是除以labels的数目，而是处以labels中不等于`ignore_index`的label的数目）。
    - `label_smoothing`：0-1之间的浮点数，用来将标签进行平滑处理，防止过拟合。如果为0，表示没有标签平滑。
    - 之前的`CrossEntropyLoss` & `NLLLoss`的输入维度是$(N, C)$或$(C)$，target维度是$(N)$或$()$，N表示样本数，C表示标签空间大小。不知从哪次更新开始，新增了一种情形，输入维度是$(N, C, d_1, d_2, ..., d_k)$，target维度是$(C, d_1, d_2, ..., d_k)$，官方文档上说，这种情形适用于2D图片逐像素计算交叉熵损失。这部分需要请cv方向的同事说一下应用场景。

### pytorch源码，[请点击这里](https://github.com/pytorch/pytorch/blob/HEAD/aten/src/ATen/native/LossNLL.cpp#L592)
- 之前看了pytorch源码，分以下几种情形：
    - 如果input和target维度相同，单独走一个分支，target必须是浮点型，计算时log_softmax(input) * target，如果`label_smoothing`不为0，target = target * (1 - label_smoothing) + label_smoothing / n_classes，然后使用上述公式计算。
    - `label_smoothing`不为0，单独一个分支
    - 其他，单独一个分支`nll_loss_nd`：根据维度，具体分为1d、2d、nd情形。

### oneflow的设计构想：
- 交叉熵函数本身不包含参数，只是一种计算方法，因此它不应该只支持单机单卡，也应当支持分布式。
- 输出根据`reduction`区分：
    - 如果为"none"，输出与输入的placement相同，sbp属性为[P, P]。
    - 如果为“sum”，输出与输入的placement相同，sbp属性变为[P, B]。
    - 如果为“mean”，输出与输入的placement相同，sbp属性变为[P, B]。
